### ISSUE TITLE
Unix.fork fails with "unlock: Operation not permitted"

### ISSUE THREAD

**Issue by kit-ty-kate** [ocamllabs/ocaml-multicore#471]
Full logs:
* http://check.ocamllabs.io:8082/log/1610933656-df43578ae336c14c63a6d491d406bb8da762395d/4.10+multicore/bad/dune.1.9.1
* http://check.ocamllabs.io:8082/log/1610933656-df43578ae336c14c63a6d491d406bb8da762395d/4.10+multicore/bad/omake.0.10.3
* http://check.ocamllabs.io:8082/log/1610933656-df43578ae336c14c63a6d491d406bb8da762395d/4.10+multicore/bad/rdf.0.11.0

All end up failing with:
```
# Fatal error during unlock: Operation not permitted
```
This is an issue with `Unix.fork` and seems to have been caused by https://github.com/ocaml-multicore/ocaml-multicore/pull/455/files#diff-f7ddfd5d7014ce34ffd8b1c4431a6e201fda084ff6b890933bef07e2f71a305a
Here is the strace:
```
clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7fc329e5ae90) = 17919
[pid 17918] fstat(4, {st_mode=S_IFIFO|0600, st_size=0, ...}) = 0
[pid 17918] lseek(4, 0, SEEK_CUR)       = -1 ESPIPE (Illegal seek)
[pid 17919] set_robust_list(0x7fc329e5aea0, 24 <unfinished ...>
[pid 17918] read(4,  <unfinished ...>
[pid 17919] <... set_robust_list resumed>) = 0
[pid 17919] write(2, "Fatal error during unlock", 25Fatal error during unlock) = 25
[pid 17919] write(2, ": Operation not permitted\n", 26: Operation not permitted
) = 26
```
_Sorry I meant to create this issue earlier but forgot._

**Comment by abbysmal** [comment]
So our intuition that it was related to 455 was founded.
The idea is that 455 is still a valid fix, however it did break the state of locks within the runtime after the fork: the domain lock is not in a valid state after the fork (in the child process), leading to any subsequent use of this lock to fail. (in this specific scenario, `Unix.close` enters a blocking section and thus releases the domain lock, failing.)
Smaller repro:

```
let () =
  let fd = Unix.dup Unix.stdout in
  let ret = Unix.fork () in
  if ret = 0 then
    Unix.close fd
  else
    print_endline "OK"
```

I have a "fix" for it but I need to think a bit to make it thorough, I'll come back with a PR once I find an appropriate solution.

**Comment by ctk21** [comment]
Closing as fixed by #491.
